README
NAME
    Map::Tube::Moscow - Interface to the Moscow Metro Map.

SYNOPSIS
     use Map::Tube::Moscow;
     my $obj = Map::Tube::Moscow->new;
     my $routes_ar = $obj->get_all_routes($from, $to);
     my $line = $obj->get_line_by_id($line_id);
     my $line = $obj->get_line_by_name($line_name);
     my $lines_ar = $obj->get_lines;
     my $station = $obj->get_node_by_id($station_id);
     my $station = $obj->get_node_by_name($station_name);
     my $route = $obj->get_shortest_route($from, $to);
     my $stations_ar = $obj->get_stations($line);
     my $metro_name = $obj->name;
     my $xml_file = $obj->xml;

DESCRIPTION
    It currently provides functionality to find the shortest route between
    the two given nodes.

    For more information about Moscow Map, click here
    <https://ru.wikipedia.org/wiki/Московский_метрополитен>.

METHODS
    "new()"
             Constructor.

    "get_all_routes($from, $to)" [EXPERIMENTAL]
             Get all routes from station to station.
             Returns reference to array with Map::Tube::Route objects.

    "get_line_by_id($line_id)"
             Get line object defined by id.
             Returns Map::Tube::Line object.

    "get_line_by_name($line_name)"
             Get line object defined by name.
             Returns Map::Tube::Line object.

    "get_lines()"
             Get lines in metro map.
             Returns reference to unsorted array with Map::Tube::Line objects.

    "get_node_by_id($station_id)"
             Get station node by id.
             Returns Map::Tube::Node object.

    "get_node_by_name($station_name)"
             Get station node by name.
             Returns Map::Tube::Node object.

    "get_shortest_route($from, $to)"
             Get shortest route between $from and $to node names. Node names in $from and $to are case insensitive.
             Returns Map::Tube::Route object.

    "get_stations($line)"
             Get list of stations for concrete metro line.
             Returns reference to array with Map::Tube::Node objects.

    "name()"
             Get metro name.
             Returns string with metro name.

    "xml()"
             Get XML specification of Moscow metro.
             Returns string with XML.

EXAMPLE1
     # Pragmas.
     use strict;
     use warnings;

     # Modules.
     use Encode qw(decode_utf8 encode_utf8);
     use Map::Tube::Moscow;

     # Object.
     my $obj = Map::Tube::Moscow->new;

     # Get route.
     my $route = $obj->get_shortest_route(decode_utf8('Планерная'), decode_utf8('Белорусская'));

     # Print out type.
     print "Route: ".encode_utf8($route)."\n";

     # Output:
     # Route: Планерная (7 Таганско-Краснопресненская линия), Сходненская (7 Таганско-Краснопресненская линия), Тушинская (7 Таганско-Краснопресненская линия), Спартак (7 Таганско-Краснопресненская линия), Щукинская (7 Таганско-Краснопресненская линия), Октябрьское поле (7 Таганско-Краснопресненская линия), Полежаевская (7 Таганско-Краснопресненская линия), Беговая (7 Таганско-Краснопресненская линия), Улица 1905 года (7 Таганско-Краснопресненская линия), Баррикадная (7 Таганско-Краснопресненская линия), Пушкинская (7 Таганско-Краснопресненская линия), Кузнецкий мост (7 Таганско-Краснопресненская линия), Китай-город (6 Калужско-Рижская линия,7 Таганско-Краснопресненская линия), Тургеневская (6 Калужско-Рижская линия), Сухаревская (6 Калужско-Рижская линия), Проспект Мира (5 Кольцевая линия,6 Калужско-Рижская линия), Новослободская (5 Кольцевая линия), Белорусская (2 Замоскворецкая линия,5 Кольцевая линия)

EXAMPLE2
     # Pragmas.
     use strict;
     use warnings;

     # Modules.
     use Map::Tube::Moscow;

     # Object.
     my $obj = Map::Tube::Moscow->new;

     # Get XML file.
     my $xml_file = $obj->xml;

     # Print out XML file.
     print "XML file: $xml_file\n";

     # Output like:
     # XML file: .*/moscow-map.xml

EXAMPLE3
     # Pragmas.
     use strict;
     use warnings;

     # Modules.
     use Map::Tube::GraphViz;
     use Map::Tube::GraphViz::Utils qw(node_color_without_label);
     use Map::Tube::Moscow;

     # Object.
     my $obj = Map::Tube::Moscow->new;

     # GraphViz object.
     my $g = Map::Tube::GraphViz->new(
             'callback_node' => \&node_color_without_label,
             'tube' => $obj,
     );

     # Get graph to file.
     $g->graph('Moscow.png');

     # Print file.
     system "ls -l Moscow.png";

     # Output like:
     # -rw-r--r-- 1 skim skim 549576 Mar  9 21:39 Moscow.png

EXAMPLE4
     # Pragmas.
     use strict;
     use warnings;

     # Modules.
     use Encode qw(encode_utf8);
     use Map::Tube::Moscow;

     # Object.
     my $obj = Map::Tube::Moscow->new;

     # Get lines.
     my $lines_ar = $obj->get_lines;

     # Print out.
     map { print encode_utf8($_->name)."\n"; } sort @{$lines_ar};

     # Output:
     # Арбатско-Покровская линия
     # Бутовская линия
     # Замоскворецкая линия
     # Калининско-Солнцевская линия
     # Калужско-Рижская линия
     # Каховская линия
     # Кольцевая линия
     # Люблинско-Дмитровская линия
     # Серпуховско-Тимирязевская линия
     # Сокольническая линия
     # Таганско-Краснопресненская линия
     # Филёвская линия

EXAMPLE5
     # Pragmas.
     use strict;
     use warnings;

     # Modules.
     use Encode qw(decode_utf8 encode_utf8);
     use Map::Tube::Moscow;

     # Arguments.
     if (@ARGV < 1) {
             print STDERR "Usage: $0 line\n";
             exit 1;
     }
     my $line = decode_utf8($ARGV[0]);

     # Object.
     my $obj = Map::Tube::Moscow->new;

     # Get stations for line.
     my $stations_ar = $obj->get_stations($line);

     # Print out.
     map { print encode_utf8($_->name)."\n"; } @{$stations_ar};

     # Output:
     # Usage: __PROG__ line

     # Output with 'foo' argument.
     # Map::Tube::get_stations(): ERROR: Invalid Line Name [foo]. (status: 105) file __PROG__ on line __LINE__

     # Output with 'Бутовская линия' argument.
     # Битцевский парк
     # Лесопарковая
     # Улица Старокачаловская
     # Улица Скобелевская
     # Бульвар адмирала Ушакова
     # Улица Горчакова
     # Бунинская аллея

DEPENDENCIES
    File::Share, Map::Tube, Moo, namespace::clean.

SEE ALSO
    Map::Tube
        Core library as Role (Moo) to process map data.

    Task::Map::Tube
        Install the Map::Tube modules.

    Task::Map::Tube::Metro
        Install the Map::Tube concrete metro modules.

REPOSITORY
    <https://github.com/tupinek/Map-Tube-Moscow>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
     © 2014-2017 Michal Josef Špaček
     Artistic License
     BSD 2-Clause License

VERSION
    0.09



Changes
0.09
 - Fix Makefile.PL for new version of Perl and no default »use lib '.'«.

0.08 2017-11-08T13:02:58+01:00
 - Add EXAMPLE5 section to doc.
 - Add test for ok_map_functions().
 - Fix color name.
 - Implement station order.
 - Rewrite XML map to new version with line IDs.
 - Update Module::Install to 1.18 version.
 - Update author name.
 - Update copyright years to 2017.
 - Update to Map::Tube 3.37.

0.07 2015-10-16T12:08:53+02:00
 - Improve description for get_lines() in doc.
 - Update to Map::Tube 3.06.

0.06 2015-10-07T13:07:12+02:00
 - Add EXAMPLE4 to doc.
 - Add get_line_by_id() and get_line_by_name() methods to doc.
 - Add test comment.
 - Add tests for get_node_by_id() and get_node_by_name() methods.
 - Fix Test::Map::Tube and planning of tests.
 - Move author tests to xt/ directory.
 - Rewrite SEE ALSO section to new style.
 - Update Module::Install to 1.16 version.

0.05 2015-03-09T21:41:26+01:00
 - Fix XML document to right map validity (Thanks Gisbert W. Selke).
 - Update images/ex3.png image for EXAMPLE3 section in doc.

0.04 2015-03-01T23:57:49+01:00
 - Add lines to XML document.
 - Add test of map validity.
 - Rewrite line stations in XML map.
 - Update map in EXAMPLE3 section in doc.

0.03 2015-02-19T08:11:55+01:00
 - Add <?xml> header to XML document.
 - Add EXAMPLE3 section to doc.
 - Add get_lines() method to doc.
 - Add get_stations() method to doc.
 - Remove executable bit from moscow-map.xml file.
 - Update SEE ALSO section in doc.
 - Update copyright years to 2015.

0.02 2014-12-09T07:54:20+01:00
 - Add Map::Tube::Berlin, Map::Tube::Minsk, Map::Tube::NYC,
   Map::Tube::Tokyo and Map::Tube::Warsaw to SEE ALSO section in doc.
 - Add Map::Tube::Kiev and Map::Tube::Sofia to SEE ALSO section in doc.
 - Add metro name.
 - Describe get_all_routes() method as EXPERIMENTAL.
 - Fix REPOSITORY section in doc.
 - Fix get_shortest_route() method description in doc.
 - Improve abstract to readable version.
 - Update LICENSE AND COPYRIGHT section in doc.

0.01 2014-12-05T07:21:25+01:00
 - First version


Makefile.PL
# Pragmas.
use lib 'inc';
use strict;
use warnings;

# Modules.
use inc::Module::Install;

# Definition.
abstract 'Interface to the Moscow Metro Map.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.08;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
install_share 'share';
license 'bsd';
license 'artistic_2';
name 'Map-Tube-Moscow';
readme_from 'Moscow.pm';
recursive_author_tests('xt');
requires 'File::Share' => 0;
requires 'Map::Tube' => '3.37';
requires 'Moo' => 0;
requires 'namespace::clean' => 0;
requires 'perl' => '5.8.0';
resources 'bugtracker' => 'https://rt.cpan.org/Public/Dist/Display.html'.
	'?Name=Map-Tube-Moscow';
resources 'homepage' => 'https://github.com/tupinek/Map-Tube-Moscow';
resources 'repository' => 'git://github.com/tupinek/Map-Tube-Moscow';
test_requires 'Encode' => 0;
test_requires 'English' => 0;
test_requires 'Test::Map::Tube' => 0.35;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.09';

# Run.
WriteAll();


Moscow.pm
package Map::Tube::Moscow;

# Pragmas.
use strict;
use warnings;
use 5.006;

# Modules.
use File::Share ':all';
use Moo;
use namespace::clean;

# Version.
our $VERSION = 0.09;

# Get XML.
has xml => (
	'is' => 'ro',
	'default' => sub {
		return dist_file('Map-Tube-Moscow', 'moscow-map.xml');
	},
);

with 'Map::Tube';

1;

__END__

=encoding utf8

=head1 NAME

Map::Tube::Moscow - Interface to the Moscow Metro Map.

=head1 SYNOPSIS

 use Map::Tube::Moscow;
 my $obj = Map::Tube::Moscow->new;
 my $routes_ar = $obj->get_all_routes($from, $to);
 my $line = $obj->get_line_by_id($line_id);
 my $line = $obj->get_line_by_name($line_name);
 my $lines_ar = $obj->get_lines;
 my $station = $obj->get_node_by_id($station_id);
 my $station = $obj->get_node_by_name($station_name);
 my $route = $obj->get_shortest_route($from, $to);
 my $stations_ar = $obj->get_stations($line);
 my $metro_name = $obj->name;
 my $xml_file = $obj->xml;

=head1 DESCRIPTION

It currently provides functionality to find the shortest route between the two
given nodes.

For more information about Moscow Map, click L<here|https://ru.wikipedia.org/wiki/Московский_метрополитен>.

=head1 METHODS

=over 8

=item C<new()>

 Constructor.

=item C<get_all_routes($from, $to)> [EXPERIMENTAL]

 Get all routes from station to station.
 Returns reference to array with Map::Tube::Route objects.

=item C<get_line_by_id($line_id)>

 Get line object defined by id.
 Returns Map::Tube::Line object.

=item C<get_line_by_name($line_name)>

 Get line object defined by name.
 Returns Map::Tube::Line object.

=item C<get_lines()>

 Get lines in metro map.
 Returns reference to unsorted array with Map::Tube::Line objects.

=item C<get_node_by_id($station_id)>

 Get station node by id.
 Returns Map::Tube::Node object.

=item C<get_node_by_name($station_name)>

 Get station node by name.
 Returns Map::Tube::Node object.

=item C<get_shortest_route($from, $to)>

 Get shortest route between $from and $to node names. Node names in $from and $to are case insensitive.
 Returns Map::Tube::Route object.

=item C<get_stations($line)>

 Get list of stations for concrete metro line.
 Returns reference to array with Map::Tube::Node objects.

=item C<name()>

 Get metro name.
 Returns string with metro name.

=item C<xml()>

 Get XML specification of Moscow metro.
 Returns string with XML.

=back

=head1 EXAMPLE1

 # Pragmas.
 use strict;
 use warnings;

 # Modules.
 use Encode qw(decode_utf8 encode_utf8);
 use Map::Tube::Moscow;

 # Object.
 my $obj = Map::Tube::Moscow->new;

 # Get route.
 my $route = $obj->get_shortest_route(decode_utf8('Планерная'), decode_utf8('Белорусская'));

 # Print out type.
 print "Route: ".encode_utf8($route)."\n";

 # Output:
 # Route: Планерная (7 Таганско-Краснопресненская линия), Сходненская (7 Таганско-Краснопресненская линия), Тушинская (7 Таганско-Краснопресненская линия), Спартак (7 Таганско-Краснопресненская линия), Щукинская (7 Таганско-Краснопресненская линия), Октябрьское поле (7 Таганско-Краснопресненская линия), Полежаевская (7 Таганско-Краснопресненская линия), Беговая (7 Таганско-Краснопресненская линия), Улица 1905 года (7 Таганско-Краснопресненская линия), Баррикадная (7 Таганско-Краснопресненская линия), Пушкинская (7 Таганско-Краснопресненская линия), Кузнецкий мост (7 Таганско-Краснопресненская линия), Китай-город (6 Калужско-Рижская линия,7 Таганско-Краснопресненская линия), Тургеневская (6 Калужско-Рижская линия), Сухаревская (6 Калужско-Рижская линия), Проспект Мира (5 Кольцевая линия,6 Калужско-Рижская линия), Новослободская (5 Кольцевая линия), Белорусская (2 Замоскворецкая линия,5 Кольцевая линия)

=head1 EXAMPLE2

 # Pragmas.
 use strict;
 use warnings;

 # Modules.
 use Map::Tube::Moscow;

 # Object.
 my $obj = Map::Tube::Moscow->new;

 # Get XML file.
 my $xml_file = $obj->xml;

 # Print out XML file.
 print "XML file: $xml_file\n";

 # Output like:
 # XML file: .*/moscow-map.xml

=head1 EXAMPLE3

 # Pragmas.
 use strict;
 use warnings;

 # Modules.
 use Map::Tube::GraphViz;
 use Map::Tube::GraphViz::Utils qw(node_color_without_label);
 use Map::Tube::Moscow;

 # Object.
 my $obj = Map::Tube::Moscow->new;

 # GraphViz object.
 my $g = Map::Tube::GraphViz->new(
         'callback_node' => \&node_color_without_label,
         'tube' => $obj,
 );

 # Get graph to file.
 $g->graph('Moscow.png');

 # Print file.
 system "ls -l Moscow.png";

 # Output like:
 # -rw-r--r-- 1 skim skim 549576 Mar  9 21:39 Moscow.png

=begin html

<a href="https://raw.githubusercontent.com/tupinek/Map-Tube-Moscow/master/images/ex3.png">
  <img src="https://raw.githubusercontent.com/tupinek/Map-Tube-Moscow/master/images/ex3.png" alt="Pražské metro" width="300px" height="300px" />
</a>

=end html

=head1 EXAMPLE4

 # Pragmas.
 use strict;
 use warnings;

 # Modules.
 use Encode qw(encode_utf8);
 use Map::Tube::Moscow;

 # Object.
 my $obj = Map::Tube::Moscow->new;

 # Get lines.
 my $lines_ar = $obj->get_lines;

 # Print out.
 map { print encode_utf8($_->name)."\n"; } sort @{$lines_ar};

 # Output:
 # Арбатско-Покровская линия
 # Бутовская линия
 # Замоскворецкая линия
 # Калининско-Солнцевская линия
 # Калужско-Рижская линия
 # Каховская линия
 # Кольцевая линия
 # Люблинско-Дмитровская линия
 # Серпуховско-Тимирязевская линия
 # Сокольническая линия
 # Таганско-Краснопресненская линия
 # Филёвская линия

=head1 EXAMPLE5

 # Pragmas.
 use strict;
 use warnings;

 # Modules.
 use Encode qw(decode_utf8 encode_utf8);
 use Map::Tube::Moscow;

 # Arguments.
 if (@ARGV < 1) {
         print STDERR "Usage: $0 line\n";
         exit 1;
 }
 my $line = decode_utf8($ARGV[0]);

 # Object.
 my $obj = Map::Tube::Moscow->new;

 # Get stations for line.
 my $stations_ar = $obj->get_stations($line);

 # Print out.
 map { print encode_utf8($_->name)."\n"; } @{$stations_ar};

 # Output:
 # Usage: __PROG__ line

 # Output with 'foo' argument.
 # Map::Tube::get_stations(): ERROR: Invalid Line Name [foo]. (status: 105) file __PROG__ on line __LINE__

 # Output with 'Бутовская линия' argument.
 # Битцевский парк
 # Лесопарковая
 # Улица Старокачаловская
 # Улица Скобелевская
 # Бульвар адмирала Ушакова
 # Улица Горчакова
 # Бунинская аллея

=head1 DEPENDENCIES

L<File::Share>,
L<Map::Tube>,
L<Moo>,
L<namespace::clean>.

=head1 SEE ALSO

=over

=item L<Map::Tube>

Core library as Role (Moo) to process map data.

=item L<Task::Map::Tube>

Install the Map::Tube modules.

=item L<Task::Map::Tube::Metro>

Install the Map::Tube concrete metro modules.

=back

=head1 REPOSITORY

L<https://github.com/tupinek/Map-Tube-Moscow>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

 © 2014-2017 Michal Josef Špaček
 Artistic License
 BSD 2-Clause License

=head1 VERSION

0.09

=cut
